content/posts/marimo.md

```diff
@@ -0,0 +1,271 @@
+[dynamical systems](https://marimo.app/l/c7h6pz). One of our community members
+even created a [QR code
+generator](https://vxlabs.com/2024/03/02/contact-qrcode-generator-with-marimo-and-wasm/).
+
+When you take a moment to think about all this, it’s really quite *magical* ✨.
+Thanks to Pyodide, you can use any pure Python package, as well as a number of
+packages for scientific computing, including Pandas, NumPy, SciPy, matplotlib,
+and scikit-learn, entirely in the sandbox of your web browser --- _without
+installing Python on your machine and without paying for infrastructure_.
+
+## Implementation: Porting marimo to Pyodide
+
+Porting marimo to work with Pyodide was relatively straightforward, a testament to the
+latter's robustness and thorough documentation.
+
+marimo's original implementation has three main parts:
+
+1. a frontend
+2. a server
+3. a backend (or "runtime") with a kernel that runs Python code.
+
+All that was required were
+minimal changes to the marimo frontend and runtime, and replacing the server
+with a lightweight bridge.
+
+### Frontend
+
+The marimo frontend uses the same build when running with Pyodide or when
+running with a Python server. This helps maintain consistency between the two
+experiences and avoid feature drift. The code path splits based off some
+feature flags that downloads the marimo wheel from PyPI and initializes the
+single-threaded, Pyodide-compatible marimo kernel. Passing only a filename and
+a message callback function, we initialize the kernel as an async, never-ending
+Python process.
+
+We interact with this kernel through a lightweight RPC bridge that sends requests to
+the kernel and receives responses through the callback passed. This allows for
+an asynchronous-like feel to the execution. To support interruption of Python
+code, we simply followed the [Pyodide
+documentation](https://pyodide.org/en/stable/usage/keyboard-interrupts.html) on
+using its `SharedArrayBuffer`-based mechanism.
+
+For performance, we cache our assets and [run Pyodide in a web
+worker](https://pyodide.org/en/stable/usage/webworker.html) to avoid blocking
+the main browser thread. This creates some extra complexity with type-safety
+and mimicking blocking RPCs.
+
+We heavily leverage the Emscripten filesystem. Code is passed from the URL hash
+(for ease of sharing), to the main frame, to the worker, and finally to the
+Emscripten filesystem. Emscripten supports many filesystem implementations such
+as IndexDB which we use to persist the user's files. We are working on our own
+implementation of the Emscripten filesystem built on top of any S3-compatible
+bucket. This means you'll be able to list, read, and write files from your S3
+bucket just by interacting with ”os” filesystem all in Python.
+
+In order to make the experience smooth in any browser environment, we do our
+best to auto-install packages. Whenever a cell is run, we try to install it
+through Pyodide or otherwise fallback to micropip.
```

Comment:
Some random questions (just curious, don't need to answer in post!)
1. Do your notebooks could lock and save the dependency set after the first run? 
2. When running locally, do you require users to manually `pip install` and manage virtual environments? Seems like this would compose better with other workflows and autoinstalling packages is risky with arbitrary install-time code execution.
3. Do you have any story for moving notebooks between Pyodide and native Python?

Reply:
```suggestion
best to auto-install packages. Whenever a cell is run, we try to install it with
`pyodide.loadPackage()` and fall back to `micropip.install()` if that fails.
```

Reply:
> Do your notebooks could lock and save the dependency set after the first run?

They don't today, but that is an excellent idea.

> When running locally, do you require users to manually pip install and manage virtual environments?

Yes, we do!

> Do you have any story for moving notebooks between Pyodide and native Python?

We haven't quite figured this out yet. In many cases no changes are required. One issue is packages — in addition to automatically installing packages, the WASM/playground frontend prompts the user to install missing packages, minimizing the need for them to explicitly use micropip. It would behoove us to have a way of shipping virtual environments or package requirements to and from Pyodide and native Python.

We would also welcome any thoughts or suggestions you might have.

Reply:
I suspect it shouldn't be too hard to lock the versions of packages used in Pyodide and then use the same versions with native Python. Going the other way is harder but I think if the user said "make sure we pick our deps so that this will work in Pyodide" then you could make a Pyodide virtual environment, pip install the packages into that, then use `pip freeze` to create a `requirements.txt` then `pip install` that into a native virtual environment.